src/plot.js

```diff
@@ -96,12 +95,38 @@ export function plot(options = {}) {
       .call(applyInlineStyles, style)
     .node();
 
-  for (const mark of marks) {
+  // A mark layout is responsible for:
+  // * applying the scales, possibly pushing a new one (for legends)
+  // * returning a filtered index and values to render, possibly new channels to rescale
+  const newScaleChannels = new Map();
+  const layouts = [];
+  const newOptions = {};
+  for (let i = 0; i < marks.length; ++i) {
+    const mark = marks[i];
+    const {channels} = layouts[i] = applyLayout.call(mark, mark.layout, [markIndex.get(mark)], scaleDescriptors, markChannels.get(mark), dimensions, options);
+    if (channels) {
+      for (const [, {scale, options: hint}] of channels) {
+        if (scale) newOptions[scale] = {...hint, ...options[scale]};
+      }
+      addScaleChannels(newScaleChannels, channels, newOptions);
+      markChannels.set(mark, channels);
+    }
+  }
+
+  // compute descriptors for new scales
+  const newScaleDescriptors = Scales(newScaleChannels, newOptions);
+  scaleDescriptors = {...newScaleDescriptors, ...scaleDescriptors};
+  scales = ScaleFunctions(scaleDescriptors);
+  for (const layout of layouts) {
+    if (layout.channels) layout.values = {...layout.values, ...applyScales(layout.channels, scales)};
+  }
+
+  for (let i = 0; i < marks.length; ++i) {
+    const mark = marks[i];
     const channels = markChannels.get(mark) ?? [];
-    const values = applyScales(channels, scales);
-    let index = markIndex.get(mark);
-    if (mark.filter != null) index = mark.filter(index, channels, values);
-    const node = mark.render(index, scales, values, dimensions, axes);
+    let {index: [I], values} = layouts[i];
```

Comment:
Given that *layout*.index is nested, we should probably call it *layout*.facets instead of *layout*.index. We make a similar distinction in that *mark*.transform returns {data, facets} (rather than {data, index}), and when the mark is not faceted, then *facets* is [*index*].

Reply:
Though it’s also the case, confusingly, that *mark*.initialize always returns {index, channels} and *index* may either be a flat array for non-faceted marks or a nested array for faceted marks.